File: lib/ipc.ts
import db from './db';

export interface IPCSection {
  section: string;
  offense: string;
  punishment: string;
  description: string;
}

export function searchIPC(query: string, limit = 3): IPCSection[] {
  // First try FTS
  const ftsStmt = db.prepare(`
    SELECT section, offense, punishment, description
    FROM ipc_fts
    WHERE ipc_fts MATCH ?
    ORDER BY rank
    LIMIT ?
  `);

  try {
    // Basic sanitization for FTS matching
    const sanitizedQuery = query.replace(/[^a-zA-Z0-9\s]/g, '').trim();
    
    if (sanitizedQuery) {
      // Create OR matched terms for robust full text searching
      const ftsQuery = sanitizedQuery.split(/\s+/).map(word => `"${word}"*`).join(' OR ');
      
      const results = ftsStmt.all(ftsQuery, limit) as IPCSection[];
      if (results.length > 0) {
        return results;
      }
    }
  } catch (error) {
    console.error('FTS query failed, falling back to LIKE', error);
  }

  // Fallback to LIKE
  const likeStmt = db.prepare(`
    SELECT section, offense, punishment, description
    FROM ipc_sections
    WHERE offense LIKE ? OR description LIKE ?
    LIMIT ?
  `);

  const likeKeyword = `%${query}%`;
  return likeStmt.all(likeKeyword, likeKeyword, limit) as IPCSection[];
}

File: backend/ipc.py
import re
from .database import get_db_connection

def search_ipc(query: str, limit: int = 3):
    conn = get_db_connection()
    cursor = conn.cursor()
    
    # Sanitize query: keep alphanumeric
    sanitized_query = re.sub(r'[^a-zA-Z0-9\s]', '', query).strip()
    
    if sanitized_query:
        # Create FTS matching syntax: "word1"* OR "word2"*
        words = sanitized_query.split()
        fts_query = " OR ".join([f'"{w}"*' for w in words])
        
        try:
            cursor.execute("""
                SELECT section, offense, punishment, description
                FROM ipc_fts
                WHERE ipc_fts MATCH ?
                ORDER BY rank
                LIMIT ?
            """, (fts_query, limit))
            
            results = cursor.fetchall()
            if results:
                return [dict(r) for r in results]
        except Exception as e:
            print("FTS match failed", e)
            
    # Fallback to LIKE
    like_query = f"%{query}%"
    cursor.execute("""
        SELECT section, offense, punishment, description
        FROM ipc_sections
        WHERE offense LIKE ? OR description LIKE ?
        LIMIT ?
    """, (like_query, like_query, limit))
    
    results = cursor.fetchall()
    return [dict(r) for r in results]
